hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/states/datanode/RunningDatanodeState.java

```diff
@@ -98,35 +148,14 @@ public void execute(ExecutorService executor) {
       }
     }
   }
-  //TODO : Cache some of these tasks instead of creating them
-  //all the time.
-  private Callable<EndpointStateMachine.EndPointStates>
-      getEndPointTask(EndpointStateMachine endpoint) {
-    switch (endpoint.getState()) {
-    case GETVERSION:
-      return new VersionEndpointTask(endpoint, conf, context.getParent()
-          .getContainer());
-    case REGISTER:
-      return  RegisterEndpointTask.newBuilder()
-          .setConfig(conf)
-          .setEndpointStateMachine(endpoint)
-          .setContext(context)
-          .setDatanodeDetails(context.getParent().getDatanodeDetails())
-          .setOzoneContainer(context.getParent().getContainer())
-          .build();
-    case HEARTBEAT:
-      return HeartbeatEndpointTask.newBuilder()
-          .setConfig(conf)
-          .setEndpointStateMachine(endpoint)
-          .setDatanodeDetails(context.getParent().getDatanodeDetails())
-          .setContext(context)
-          .build();
-    case SHUTDOWN:
-      break;
-    default:
-      throw new IllegalArgumentException("Illegal Argument.");
+
+  private Callable<EndPointStates> getEndPointTask(
+      EndpointStateMachine endpoint) {
+    if (endpointTasks.containsKey(endpoint.toString())) {
+      return endpointTasks.get(endpoint.toString()).get(endpoint.getState());
```

Comment:
Can you please explain why a string-based map is better than using the endpoint itself as a key?

Reply:
I just wan to use a more simple identity to indicate the different  Endpoint instance. So I use the endpoint address. But I have updated this to use Endpoint itself now.